### PR TITLE
Add a suggestion to replace `map(f).unwrap_or(None)` with `and_then(f)`.

### LINKED ISSUE #2109: Suggest `.and_then(f)` instead of `.map(f).unwrap_or(None)`
Recently I've come accross this code in one of my projects:

```rust
get_some_option_value()
    .map(some_calculation_that_returns_option)
    .unwrap_or(None)
```

Since I wrote that, I learned a bit more about Option's api so I changed it to `.and_then`.

This is the kind of stuff I would expect clippy to tell me about, so I suggest adding a lint to catch this.

### PR DESCRIPTION
**Pull request by topecongiro**
Closes #2109.

### CODE REVIEW

File: clippy_lints/src/methods.rs [hunk 1]
```diff
@@ -1150,29 +1150,35 @@ fn lint_ok_expect(cx: &LateContext, expr: &hir::Expr, ok_args: &[hir::Expr]) {
...
     // lint if the caller of `map()` is an `Option`
     if match_type(cx, cx.tables.expr_ty(&map_args[0]), &paths::OPTION) {
-        // lint message
-        let msg = "called `map(f).unwrap_or(a)` on an Option value. This can be done more directly by calling \
-                   `map_or(a, f)` instead";
         // get snippets for args to map() and unwrap_or()
         let map_snippet = snippet(cx, map_args[1].span, "..");
         let unwrap_snippet = snippet(cx, unwrap_args[1].span, "..");
+        // lint message
+        let arg = if unwrap_snippet == "None" { "None" } else { "a" };
```
[the remark this answers is not in the document]
Comment (topecongiro) [reply clippy_lints/src/methods.rs]: Hey sorry for horrifying you - I had no hesitant on comparing snippets to raw text 😛

[the remark this answers is not in the document]
Comment (topecongiro) [reply clippy_lints/src/methods.rs]: While adding more tests, I found that comparing snippets to raw text does not work in some trivial cases. I copied them from tests for `OPTION_MAP_UNWRAP_OR`. e.g.
```rust
    let _ = opt.map(|x| Some(x + 1))
        .unwrap_or({
            None
        });
```
Is this acceptable, or should we "do it right"?

[the remark this answers is not in the document]
Comment (oli-obk) [reply clippy_lints/src/methods.rs]: Yea that's fine

**Comment by topecongiro** [comment]
Added a comment and tests.

**Comment by oli-obk** [comment]
Yay! So... now that this one is through, I noticed the lint (the original one already had this issue) does not have a structured suggestion (`span_suggestion`). If you are looking for something to do, this would be something fairly easy, plus you already know the code here. Adding structured suggestions will allow RLS to offer to fix the issue with a click instead of having to do it manually

**Pull request merged**